README.md



# linker

A flutter plugin to link with other app or system function.

## Getting Started

add 

```

linker: 0.0.2

```
to your pubspec.yaml, and run `flutter packages get`


## Usage

### Android


Use action

```
try {
  ActivityResult result = await Linker.startActivityForResult(
      new Intent.fromAction("android.settings.SETTINGS"), 0);
  print(result);
} on PlatformException catch (e) {
  print("Open failed $e");
}

```

Open google map

```
try {
  await Linker.startActivity(new Intent.fromAction(
    "android.intent.action.VIEW",
    uri: Uri.parse(
        "http://ditu.google.cn/maps?hl=zh&mrt=loc&q=31.1198723,121.1099877(上海青浦大街100号)"),
  ));
} on PlatformException catch (e) {
  print("Open failed $e");
}

```


Open QQ

```
 try {
      await Linker.startActivity(new Intent.fromAction(Intent.ACTION_VIEW,
          uri: Uri.parse("mqqwpa://im/chat?chat_type=wpa&uin=123456")));
    } on PlatformException catch (e) {
      print("Open failed $e");
    }
```



Open Wechat
```
try {
  await Linker.startActivity(new Intent.callApp( packageName:  "com.tencent.mm",className: "com.tencent.mm.ui.LauncherUI"));
} on PlatformException catch (e) {
  print("Open failed $e");
}
```

### ios


Open wechat

```
 try {
 await Linker.openURL("weixin://");
} on PlatformException catch (e) {
  print("Open failed $e");
}
```


Open Settings
```
try {
  await Linker.openURL("App-Prefs:root");
} on PlatformException catch (e) {
  print("Open failed $e");
}

```

Call 10086

```
 try {
  await Linker.openURL("tel:10086");
} on PlatformException catch (e) {
  print("Open failed $e");
}
```











CHANGELOG.md
## 0.0.2
    * support Intent.ACTION_VIEW in android
    * add open qq/wechat in android

## 0.0.1
    * ios canOpenURL
    * ios openURL
    * android startActivityForResult
    * android startActivity


example/test/widget_test.dart
// This is a basic Flutter widget test.
// To perform an interaction with a widget in your test, use the WidgetTester utility that Flutter
// provides. For example, you can send tap and scroll gestures. You can also use WidgetTester to
// find child widgets in the widget tree, read text, and verify that the values of widget properties
// are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:linker_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(new MyApp());

    // Verify that platform version is retrieved.
    expect(
        find.byWidgetPredicate(
          (Widget widget) =>
              widget is Text && widget.data.startsWith('Running on:'),
        ),
        findsOneWidget);
  });
}


lib/linker.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter/services.dart';
import 'package:meta/meta.dart';

@immutable
class Intent {


  static const String ACTION_WIRELESS_SETTINGS="android.settings.WIRELESS_SETTINGS";
  static const String ACTION_SETTINGS="android.settings.SETTINGS";
  static const String ACTION_VIEW="android.intent.action.VIEW";


  final String className;
  final String packageName;
  final String action;
  final Uri uri;
  final Map<String, dynamic> extras;

  Intent._(
      {this.className, this.action, this.packageName, this.uri, this.extras});

  factory Intent.fromAction(String action,
      {Uri uri, String packageName, String className}) {
    return new Intent._(
        action: action,
        uri: uri,
        className: className,
        packageName: packageName);
  }

  factory Intent.callApp(
      {String className, String packageName, Map<String, dynamic> extras}) {
    return new Intent._(
        packageName: packageName, className: className, extras: extras);
  }

  Map toMap() {
    return {
      "className": className,
      "packageName": packageName,
      "extras": extras,
      "action": action,
      "uri": uri?.toString(),
    };
  }
}

class ActivityResult {
  static const int RESULT_CANCELED = 0;
  static const int RESULT_FIRST_USER = 1;
  static const int RESULT_OK = -1;

  ///
  final int resultCode;

  ///
  final int requestCode;

  ///
  final Intent intent;

  bool get isOk {
    return resultCode == RESULT_OK;
  }

  bool get isCancel {
    return resultCode == RESULT_CANCELED;
  }

  ActivityResult._({this.requestCode, this.resultCode, this.intent});

  factory ActivityResult.fromMap(dynamic map) {
    dynamic _intent = map['intent'];
    Intent intent;
    if (_intent != null) {
      intent = new Intent._(
        action: _intent['action'],
      );
    }
    return new ActivityResult._(
        requestCode: map['requestCode'],
        resultCode: map['resultCode'],
        intent: intent);
  }

  @override
  String toString() {
    return "ActivityResult {requestCode:$requestCode,resultCode:$resultCode,intent:$intent}";
  }
}

class Linker {
  static const MethodChannel _channel = const MethodChannel('linker');

  /// IOS only
  static Future<bool> canOpenURL(String url) async {
    if (!Platform.isIOS)
      throw new Exception("This method must be called in ios");
    dynamic value = await _channel.invokeMethod("canOpenURL", url);
    return value as bool;
  }

  /// IOS only,
  static Future<bool> openURL(String url) async {
    if (!Platform.isIOS)
      throw new Exception("This method must be called in ios");
    dynamic value = await _channel.invokeMethod("openURL", url);
    return value as bool;
  }

  /// android only
  static Future<ActivityResult> startActivityForResult(
      Intent intent, int requestCode) async {
    if (!Platform.isAndroid)
      throw new Exception("This method must be called in android");
    Map data = intent.toMap();
    data['requestCode'] = requestCode;
    dynamic value = await _channel.invokeMethod("startActivityForResult", data);
    return new ActivityResult.fromMap(value);
  }

  /// android only
  static Future<bool> startActivity(Intent intent) async {
    if (!Platform.isAndroid)
      throw new Exception("This method must be called in android");
    dynamic value =
        await _channel.invokeMethod("startActivity", intent.toMap());
    return value as bool;
  }

  static Future<bool> openSetting() async{
    var value = await _channel.invokeMethod("openSetting");
    return value as bool;
  }
  static Future<bool> openNetworkSetting() async{
    if(Platform.isAndroid){
      return startActivity(new Intent.fromAction(Intent.ACTION_WIRELESS_SETTINGS));
    }else{
      return openSetting();
    }
  }
}


ios/Classes/LinkerPlugin.m
#import "LinkerPlugin.h"


@interface Promise : NSObject

-(id) get;

-(void)set:(id)value;

@end


@interface Promise()
{
    id _value;
    NSCondition *_lock;
}

@end


@implementation Promise

-(id)init{
    if (self = [super init]){
        
        _lock = [[NSCondition alloc]init];
    }
    return self;
}


-(id)get{
    if(_value){
        return _value;
    }
    
    [_lock lock];
    while(!_value){
        [_lock wait];
    }
    [_lock unlock];
    return _value;
}

-(void)set:(id)value{
    [_lock lock];
    _value = value;
    [_lock signal];
    [_lock unlock];
}

@end


@interface LinkerPlugin()

@property (nonatomic,weak) FlutterMethodChannel* channel;

@end

@implementation LinkerPlugin



+(LinkerPlugin*)sharedInstance{
    static LinkerPlugin *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[LinkerPlugin alloc]init];
    });
    return sharedInstance;
}


+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"linker"
            binaryMessenger:[registrar messenger]];
    LinkerPlugin* instance = [LinkerPlugin sharedInstance];
    instance.channel = channel;
  [registrar addMethodCallDelegate:instance channel:channel];
}




-(BOOL)handleOpenURL:(NSURL *)url{
    Promise* promise = [[Promise alloc]init];
    [_channel invokeMethod:@"handleOpenURL" arguments:url.absoluteString result:^(id  _Nullable result) {
        [promise set:result];
    }];
    
    return [[promise get]boolValue];
    
}



- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* method = call.method;
    
  if ([@"openURL" isEqualToString:method]) {
      if (@available(iOS 10.0, *)) {
          [[UIApplication sharedApplication]openURL:[NSURL URLWithString:call.arguments] options:@{} completionHandler:^(BOOL success) {
              result(@(success));
          }];
      } else {
          BOOL success = [[UIApplication sharedApplication]openURL:[NSURL URLWithString:call.arguments]];
          result(@(success));
      }
  }else if([@"canOpenURL" isEqualToString:method]){
      
      BOOL value = [[UIApplication sharedApplication]canOpenURL:[NSURL URLWithString:call.arguments]];
      result(@(value));
  }else if([@"openSetting" isEqualToString:method]){
      NSURL * url = [NSURL URLWithString:UIApplicationOpenSettingsURLString];
      if([[UIApplication sharedApplication] canOpenURL:url]) {
          NSURL*url =[NSURL URLWithString:UIApplicationOpenSettingsURLString];
         BOOL value =  [[UIApplication sharedApplication] openURL:url];
          result(@(value));
      }else{
          result(@(NO));
      }
  }  else {
    result(FlutterMethodNotImplemented);
  }
}

@end


android/src/main/java/com/github/jzoom/linker/LinkerPlugin.java
package com.github.jzoom.linker;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** LinkerPlugin */
public class LinkerPlugin implements MethodCallHandler, PluginRegistry.ActivityResultListener {

  final MethodChannel channel;
  final Activity activity;

  public LinkerPlugin(MethodChannel channel, Activity activity) {
    this.channel = channel;
    this.activity = activity;


  }

  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "linker");
    LinkerPlugin linkerPlugin = new LinkerPlugin(channel,registrar.activity());
    registrar.addActivityResultListener(linkerPlugin);
    channel.setMethodCallHandler(linkerPlugin);
  }

  private Result result;

  private int requestCode;


  static void setValueToIntent(String name, Object value, Intent map){
    if(value instanceof String) {
      map.putExtra(name, (String)value);
    } else if(value instanceof Integer) {
      map.putExtra(name, ((Integer)value).intValue());
    } else if(value instanceof Float) {
      map.putExtra(name, ((Float)value).floatValue());
    } else if(value instanceof Double) {
      map.putExtra(name, ((Double)value).doubleValue());
    } else if(value instanceof Boolean) {
      map.putExtra(name, ((Boolean)value).booleanValue());
    }else{
      throw new RuntimeException("Unsupported type:" + value.getClass());
    }
  }

  Map<String,Object> bundle2Map(Bundle bundle){
    if(bundle==null){
      return null;
    }
    Map<String,Object> data = new HashMap<>();

    Set<String> set = bundle.keySet();
    for (String key:set) {
      /// Notice the value must be double/int/boolean/String
      data.put(key,bundle.get(key));
    }

    return data;

  }
  private Map intent2map(Intent intent) {
    Map map = new HashMap();
    map.put("action",intent);
    map.put("extras",intent.getExtras());


    return map;

  }
  private Intent parseIntent( Map data ){

    Map<String,Object> extras = (Map<String,Object>) data.get("extras");
    Intent intent = null;
    String uri = (String) data.get("uri");
   String action = (String) data.get("action");

    if(action!=null){
      if(uri!=null){
        intent = new Intent(action, Uri.parse(uri));
      }else{
        intent = new Intent(action);
      }
    }

    if(intent==null){
      intent = new Intent();
    }

    String packageName = (String) data.get("packageName");
    String className = (String) data.get("className");
    if(packageName!=null && className!=null){
      intent.setClassName(packageName, className);
    }

    if(extras!=null){
      for(Map.Entry<String,Object> entry : extras.entrySet()){
        String key = entry.getKey();
        Object value = entry.getValue();
        if(value==null){
          continue;
        }
        setValueToIntent(key,value,intent);
      }
    }


    return intent;

  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    String method = call.method;
    if ("startActivityForResult".equals(method)) {
      synchronized (this){
        if(this.result!=null){
          this.result.error("cancel","The last result is not finished","");
          this.result = null;
        }

        Map data = (Map) call.arguments;
        this.result = result;
        this.requestCode = (Integer)data.get("requestCode");
        Intent intent = parseIntent(data);
        try{
          activity.startActivityForResult(intent,requestCode);
        }catch (Throwable e){
          result.error(e.getClass().getName(),e.getMessage(),"Cannot start activity");
          this.result = null;
        }
      }
    }else if("startActivity".equals(method)){
      Map data = (Map) call.arguments;
      Intent intent = parseIntent(data);
      try{
        activity.startActivity(intent);
        result.success(true);
      }catch (Throwable e){
        result.error(e.getClass().getName(),e.getMessage(),"Cannot start activity");
      }

    }else if("openSetting".equals(method)){
      Intent intent = new Intent();
      intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
      if (Build.VERSION.SDK_INT >= 9) {
        intent.setAction("android.settings.APPLICATION_DETAILS_SETTINGS");
        intent.setData(Uri.fromParts("package", getPackageName(), null));
      } else if (Build.VERSION.SDK_INT <= 8) {
        intent.setAction(Intent.ACTION_VIEW);
        intent.setClassName("com.android.settings","com.android.settings.InstalledAppDetails");
        intent.putExtra("com.android.settings.ApplicationPkgName", getPackageName());
      }
      activity.startActivity(intent);
      result.success(true);
    }  else {
      result.notImplemented();
    }
  }

  private String getPackageName() {
    return activity.getPackageName();
  }


  @Override
  public boolean onActivityResult(int requestCode, int resultCode, Intent intent) {
    synchronized (this){
      if(requestCode == this.requestCode && result!=null){
        Map<String,Object> data = new HashMap<>();
        data.put("resultCode",resultCode);
        data.put("requestCode",requestCode);
        if(intent!=null){
          data.put("intent", intent2map(intent) );
        }


        result.success(data);
        this.result = null;
        return true;
      }
    }
    return false;
  }


}


example/lib/main.dart
import 'dart:io';

import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:linker/linker.dart';

void main() => runApp(new MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => new _MyAppState();
}

class Button extends StatelessWidget {
  final String label;

  final VoidCallback onPressed;

  Button({this.label, this.onPressed});

  @override
  Widget build(BuildContext context) {
    return new RaisedButton(
        onPressed: onPressed,
        color: Colors.blueAccent,
        textColor: Colors.white,
        child: new Text(label));
  }
}

class AndroidExample extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new Center(
      child: new Column(
        mainAxisSize: MainAxisSize.min,
        children: <Widget>[
          new Button(
              onPressed: () async {
                try {
                  await Linker.openSetting();
                } on PlatformException catch (e) {
                  print("Open failed $e");
                }
              },
              label: "Open Settings"),
          new Button(
              onPressed: () async {
                try {
                  await Linker.startActivity(new Intent.fromAction(Intent.ACTION_VIEW,
                      uri: Uri.parse("mqqwpa://im/chat?chat_type=wpa&uin=123456")));
                } on PlatformException catch (e) {
                  print("Open failed $e");
                }
              },
              label: "Open QQ"),
          new Button(
              onPressed: () async {
                try {
                  await Linker.startActivity(new Intent.callApp( packageName:  "com.tencent.mm",className: "com.tencent.mm.ui.LauncherUI"));
                } on PlatformException catch (e) {
                  print("Open failed $e");
                }
              },
              label: "Open Wechat"),
          new Button(
              onPressed: () async {
                try {
                  await Linker.openNetworkSetting();
                } on PlatformException catch (e) {
                  print("Open failed $e");
                }
              },
              label: "Open Network settings"),
          new Button(
              onPressed: () async {
                try {
                  await Linker.startActivity(new Intent.fromAction(
                    "android.intent.action.VIEW",
                    uri: Uri.parse(
                        "http://ditu.google.cn/maps?hl=zh&mrt=loc&q=31.1198723,121.1099877(上海青浦大街100号)"),
                  ));
                } on PlatformException catch (e) {
                  print("Open failed $e");
                }
              },
              label: "Open google map"),
        ],
      ),
    );
  }
}

class IosExample extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new Center(
      child: new Column(
        mainAxisSize: MainAxisSize.min,
        children: <Widget>[
          new Button(
              onPressed: () async {
                try {
                  await Linker.openURL("weixin://");
                } on PlatformException catch (e) {
                  print("Open failed $e");
                }
              },
              label: "Open Wechat"),
          new Button(
              onPressed: () async {
                try {
                  await Linker.openSetting();
                } on PlatformException catch (e) {
                  print("Open failed $e");
                }
              },
              label: "Open Settings"),
          new Button(
              onPressed: () async {
                try {
                  await Linker.openURL("tel:10086");
                } on PlatformException catch (e) {
                  print("Open failed $e");
                }
              },
              label: "Call 10086"),
        ],
      ),
    );
  }
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      home: new Scaffold(
        appBar: new AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Platform.isAndroid ? new AndroidExample() : new IosExample(),
      ),
    );
  }
}
